tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php

```diff
@@ -0,0 +1,277 @@
+        [$author1, $author2, $author3] = $this->createTestData();
+     // all authors with at least one good book according to GoodBookQuery
+        $authors = AuthorQuery::create()
+        ->useExistsQuery('Book', null, GoodBookQuery::class)
+        ->filterByIsGood()
+        ->endUse()
+        ->find($this->con)
+        ->getData();
+
+        $this->assertEquals([$author1, $author3], $authors);
+    }
+
+    /**
+     * @return void
+     */
+    public function testUseExistsQueryCanNestUseQuery()
+    {
+        [$author1, $author2, $author3] = $this->createTestData();
+        (new Essay())->setFirstAuthor($author1)->setTitle('leEssay')->save($this->con);
+        (new Essay())->setFirstAuthor($author1)->setTitle('leEssay')->save($this->con);
+        (new Essay())->setFirstAuthor($author3)->setTitle('leEssay')->save($this->con);
+
+        // all books of authors who are first author of an essay called 'leEssay'
+        $books = BookQuery::create()
+        ->useExistsQuery('Author')
+        ->useEssayRelatedByFirstAuthorIdQuery()
+        ->filterByTitle('leEssay')
+        ->endUse()
+        ->endUse()
+        ->find($this->con)
+        ->getData();
+
+        $this->assertCount($author1->countBooks(null, false, $this->con) + $author3->countBooks(null, false, $this->con), $books);
+    }
+
+    /**
+     * @return void
+     */
+    public function testUseExistsQueryDoesNotMergeWithOuterQuery()
+    {
+        $sawException = false;
+        $innerBookQuery = ExceptionOnMergeAuthorQuery::create()->useExistsQuery('Book');
+        try {
+            $innerBookQuery->endUse();
+        } catch (Exception $e) {
+            $sawException = true;
+        }
+        $this->assertFalse($sawException, 'endUse() should not call mergeWidth');
+    }
+
+    /**
+     * Creates and returns four authors, one and three have a good book, one and two have a bad book, four has no book
+     *
+     * @return \Propel\Tests\Bookstore\Author[]
+     */
+    private function createTestData()
+    {
+        BookQuery::create()->deleteAll($this->con);
+        AuthorQuery::create()->deleteAll($this->con);
+
+        $author1 = $this->createAuthor('LeAuthor1');
+        $this->createBook($author1);
+        $this->createBook($author1, 1);
+        $this->createBook($author1, 1);
+
+        $author2 = $this->createAuthor('LeAuthor2');
+        $this->createBook($author2);
+        $this->createBook($author2);
+
+        $author3 = $this->createAuthor('LeAuthor3');
+        $this->createBook($author3, 1);
+
+        $author4 = $this->createAuthor('LeAuthor4');
+
+        return [$author1, $author2, $author3, $author4];
+    }
+
+    private function createAuthor(string $lastName, string $firstname = 'LeDefaultFirstname')
+    {
+        $author = new Author();
+        $author->setLastName($lastName)->setFirstName($firstname)->save($this->con);
+
+        return $author;
+    }
+
+    private function createBook(Author $author, $isGood = false)
```

Comment:
A very minor remark: the test would tell its 'story' better, when the `$isGood` boolean parameter would be replaced with a `string $title`

Reply:
That would be more readable by itself, yes. In this particular case, I honestly like it better this way, as it makes the `createTestData()` method less cluttered, and allows you to see the whole data setup in just a glance, which was helpful when writing the tests. When you know what to look for, you can identify the number of "good" books in about a second.

Now that I have written the tests and don't rely on the short glace reminder anymore, I don't mind changing it to a more verbose signature, if you insist

Reply:
The PHP interpreter doesn't mind either way, but it will save an interested developer/maintainer in a few months some clicks and thinking when scanning `createTestData`. In general I would say it's worth the effort to strive for maximum clarity